### PR TITLE
Permission Handling

### PR DESCRIPTION
**Pull request by ControlTheBit**
I am creating this PR so that we can properly track the progress of #6905.

**Missing:**
Status: _DO NOT MERGE_

**Explanation of what this change does**
This change allows c:geo to handle location permissions on demand.
What is does (pseudocode):

- if app.hasLocationPermissions
    - work like normal
- else 
    - Ask user for permission
        - if user denies permissions
           - Show popup where the user can decide to be asked again or close the app
        - else 
            - work like normal

### CODE REVIEW

**Comment by Lineflyer** [comment]
Ok to test

**Comment by Lineflyer** [comment]
@ControlTheBit Thanks very much for your PR and the effort. The description of the working mode is IMHO what we need right now. I am not able to review the code itself due to lack of coding experience, but some comments from my side:
- Translation wil be taken care of via crowdin as soon as your PR gets merged.
- The build result is failed with one unrelated failed test (Geocoder) but seems two other tests also need adaption according to the build result

Besides that I volunteer for testing either after merge (if other team members are OK to merge) or I can also test the version including your PR if needed. Do you of special situations which might need to be tested or "just" a explorative test?

**Comment by rsudev** [comment]
I'll try to take a look into the code tomorrow

**Comment by Lineflyer** [comment]
@rsudev Highly apprecciated!!

**Comment by ControlTheBit** [comment]
I had a brief look at the three failing tests yesterday and will try to fix them this week before Friday unless someone else finds the time to have a look. I think it has to do with the permissions that are probably missing in the tests.

If that is the case, then it should be enough to use the [GrantPermissionRule class](https://developer.android.com/reference/android/support/test/rule/GrantPermissionRule) to grant the needed permissions while executing the tests. The needed permissions should be `ACCESS_FINE_LOCATION` and `ACCESS_COARSE_LOCATION`.

File: main/src/cgeo/geocaching/sensors/GeoData.java [hunk 1]
```diff
@@ -91,7 +100,6 @@ public static GeoData getInitialLocation(final Context context) {
         final String homeLocationStr = Settings.getHomeLocation();
         if (StringUtils.isNotBlank(homeLocationStr)) {
             try {
-                assert homeLocationStr != null;
```
[the remark this answers is not in the document]
Comment (ControlTheBit) [reply main/src/cgeo/geocaching/sensors/GeoData.java]: Bacause afaik asserts are for nothing unless they are explicitly actived in the JVM

**Comment by ControlTheBit** [comment]
@rsudev thanks for looking over my code :-) I see the problem of formatting, I should not have committed that to this PR and I'm not sure how we can get into this "GPS off" mode.

The current way of handling with the permission was discussed in #6905. I think this is "ok" for a first version to just close the app. The app must be able to be published to the app store in case geocaching.com makes sudden changes or something. We will not be able to publish any updates after Nov 1st (I think) if we do not target at least sdk 26.

I know this is very bad timing to say, but I can't take care of this right now. As I already said in #6905, I got a new project at work and am quite busy with that. I hope someone of you can find the time to have a deeper look and implement missing pieces if they need to be in the first version of this feature. Please don't hesitate so let me know if something in the code is unclear. I will try to make some time next week, but can't promise anything right now. :-(

File: main/src/cgeo/geocaching/sensors/GeoData.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (UniQP) [reply main/src/cgeo/geocaching/sensors/GeoData.java]: IIRC the main purpose of these `assert foo != null;` statements is to silence some Eclipse warnings.

**Comment by UniQP** [comment]
I will clean up the commits and integrate the suggestions by @rsudev.

**Comment by Lineflyer** [comment]
@rsudev @ControlTheBit Do you consider this ready for testing in the nightlies?

**Comment by Lineflyer** [comment]
retest this please

**Comment by Lineflyer** [comment]
@ControlTheBit Are you planning to do changes on this PR or else (as suggested by @rsudev ) he suggested to optimize the open points later.
@rsudev What about merging this one? There might be only the problem of the failing tests for our CI?!

**Comment by rsudev** [comment]
Just as a heads-up, I started to cross check again as announced. Right now I am a bit confused by some behaviours I see and in addition I thought it did initially also check for the storage permission? The app works without that, but uses the internal or external data folder, even for the backup, which is kind of counter-productive. Need to compare some more things in comparison to the non-target-sdk-26 version.

**Comment by rsudev** [comment]
One thing to note (which I will look into) - the new version seem to leak a MainActivity on 'closing'.

**Comment by Lineflyer** [comment]
I installed the debug build from CI now on an Android 8 tablet device to give it a try.
As Idon't know what is intended to work and what issues are still known to be solved, this is what I have seen while playing around:
- Location permission request seems to work good so far
- I did not manage to trigger the storage permission request  
- Clicking on "Live Map" crashes the app
- The design of the action bar changed and it misses the title on some screens (main screen, Cache details)
- There seem to be issues with design colors as well...e.g. the trackable inventory of a cache looks broken (black on black, icons not in same row with TBs)

If this is all known or not yet intended to work, just dismiss my list :)

**Comment by ControlTheBit** [comment]
@Lineflyer for me it is totally fine if @rsudev can take care of the remaining points. :-)

**Comment by Lineflyer** [comment]
@Bananeweizen 
From the code history and AFAIR you have been involved or even triggered the last changes of minSDK and targetSDK in c:geo. We currently have this PR pending, which is desperately needed as GooglePlay only accepts raised targetSDK as of now. If you find time and are interested. Its really needed to keep us alive.

**Comment by rsudev** [comment]
As discussed on IRC some days ago, the Google map crashes reproducible with this change. The reason is, that the activity doesn't have an ActionBar any more. While the crash can be easily mitigated did I not find a way yet to put the ActionBar back in when looking into the problem yesterday.

**Comment by Lineflyer** [comment]
@ControlTheBit 
First of all: Thanks for getting us started with the permission handling topic and providing a valuable foundation for the respective code change. This PR was meanwhile superseded by #7213, only therefore I am closing this PR.

You are more than welcome to further contribute and/or help with this or also other topics!!

**Pull request closed without merging**